test/functions.jl

```diff
@@ -0,0 +1,152 @@
+@testset "map" begin
+    ds = KeyedDataset(
+        flatten([
+            :g1 => [
+                :a => KeyedArray(zeros(3); time=1:3),
+                :b => KeyedArray(ones(3, 2); time=1:3, loc=[:x, :y]),
+            ],
+            :g2 => [
+                :a => KeyedArray(ones(3); time=1:3),
+                :b => KeyedArray(zeros(3, 2); time=1:3, loc=[:x, :y]),
+            ]
+        ])...
+    );
+
+    # Since all of the tests are reusing `ds` we can assum it isn't being mutated.
+    # Maybe there's a better way to test the combinations, but this seemed pretty terse?
+
+    # Test mapping over every component
+    r = map(a -> a .+ 1.0, ds)
+    @test mean(r[(:g1, :a)]) == mean(r[(:g2, :b)]) == 1.0
+    @test mean(r[(:g1, :b)]) == mean(r[(:g2, :a)]) == 2.0
+
+    # Test mapping over components
+    r = map(a -> a .+ 1.0, ds, (:g1, :__))
+    @test mean(r[(:g2, :b)]) == 0.0
+    @test mean(r[(:g1, :a)]) == mean(r[(:g2, :a)]) == 1.0
+    @test mean(r[(:g1, :b)]) == 2.0
+
+    # Test mapping over more complex patterns
+    r = map(a -> a .+ 1.0, ds, (:__, :a, :_))
+    @test mean(r[(:g2, :b)]) == 0.0
+    @test mean(r[(:g1, :a)]) == mean(r[(:g1, :b)]) == 1.0
+    @test mean(r[(:g2, :a)]) == 2.0
+
+    # Test mapping just dims
+    r = map(a -> a .+ 1.0, ds, (:__, :loc))
+    @test mean(r[(:g1, :a)]) == 0.0
+    @test mean(r[(:g2, :b)]) == mean(r[(:g2, :a)]) == 1.0
+    @test mean(r[(:g1, :b)]) == 2.0
+
+    # Test mapping over component and dim pattern
+    r = map(a -> a .+ 1.0, ds, (:g1, :__, :loc))
+    @test mean(r[(:g1, :a)]) == mean(r[(:g2, :b)]) == 0.0
+    @test mean(r[(:g2, :a)]) == 1.0
+    @test mean(r[(:g1, :b)]) == 2.0
+end
+
+@testset "mapslices" begin
+    ds = KeyedDataset(
+        :val1 => KeyedArray(zeros(3, 2); time=1:3, obj=[:a, :b]),
+        :val2 => KeyedArray(ones(3, 2); time=1:3, loc=[:x, :y]),
+    );
+
+    r = mapslices(sum, ds; dims=:time);
```

Comment:
It would probably be a good idea to include a test using `keys` as well.

Reply:
Added a few more tests for that and also some error cases (e.g., dimension doesn't exist, constraint violated).